examples/request-id/src/main.rs

```diff
@@ -0,0 +1,92 @@
+//! Run with
+//!
+//! ```not_rust
+//! cargo run -p example-request-id
+//! ```
+
+use axum::{
+    extract::MatchedPath,
+    http::{HeaderName, Request},
+    response::Html,
+    routing::get,
+    Router,
+};
+use tower::ServiceBuilder;
+use tower_http::{
+    request_id::{MakeRequestUuid, PropagateRequestIdLayer, SetRequestIdLayer},
+    trace::TraceLayer,
+};
+use tracing::{info, info_span};
+use tracing_subscriber::{layer::SubscriberExt, util::SubscriberInitExt};
+
+const REQUEST_ID_HEADER: &str = "x-request-id";
+
+#[tokio::main]
+async fn main() {
+    tracing_subscriber::registry()
+        .with(
+            tracing_subscriber::EnvFilter::try_from_default_env().unwrap_or_else(|_| {
+                // axum logs rejections from built-in extractors with the `axum::rejection`
+                // target, at `TRACE` level. `axum::rejection=trace` enables showing those events
+                format!(
+                    "{}=debug,tower_http=debug,axum::rejection=trace",
+                    env!("CARGO_CRATE_NAME")
+                )
+                .into()
+            }),
+        )
+        .with(tracing_subscriber::fmt::layer())
+        .init();
+
+    let x_request_id = HeaderName::from_static(REQUEST_ID_HEADER);
+
+    let middleware = ServiceBuilder::new()
+        .layer(SetRequestIdLayer::new(
+            x_request_id.clone(),
+            MakeRequestUuid,
+        ))
+        .layer(
+            TraceLayer::new_for_http().make_span_with(|request: &Request<_>| {
+                // Log the matched route's path (with placeholders not filled in).
+                // Use request.uri() or OriginalUri if you want the real path.
+                let matched_path = request
+                    .extensions()
+                    .get::<MatchedPath>()
+                    .map(MatchedPath::as_str);
+
+                // Log the request id as generated.
+                let request_id = request.headers().get(REQUEST_ID_HEADER);
+
+                match request_id {
+                    Some(request_id) => info_span!(
+                        "http_request",
+                        matched_path,
+                        method = ?request.method(),
+                        request_id = ?request_id,
+                    ),
+                    None => info_span!(
```

Comment:
This should never be actually hit, right? If so, I think it would be better to either place a comment here saying as much or using `tracing::error!` or similar to signal that this would be a bug.

Right now it seems to me like this expects to sometimes not have the request ID.

Reply:
Good call! I've updated `None` case to log an error.